core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -901,6 +900,67 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
   }
 
+  def handleCreateTopicRequest(request: RequestChannel.Request) {
+    val createTopicRequest = request.body.asInstanceOf[CreateTopicRequest]
+
+    def handleUnauthorizedRequest = {
+      createTopicRequest.topics.asScala.map { case (topic, _) =>
+        (topic, Errors.CLUSTER_AUTHORIZATION_FAILED)
+      }
+    }
+
+    def handleAuthorizedRequest = {
+      val brokers = metadataCache.getAliveBrokers.map(_.id)
+      createTopicRequest.topics.asScala.map { case (topic, arguments) =>
+        try {
+          // Prep configs
+          val configs = new Properties()
+          arguments.configs.asScala.foreach { case (key, value) =>
+            configs.setProperty(key, value)
+          }
+          LogConfig.validate(configs)
+
+          // Prep assignment
+          val assignments = {
+            if (!arguments.replicasAssignments.isEmpty) {
+              // Note: we don't check that replicaAssignment doesn't contain unknown brokers - unlike in add-partitions case,
+              // this follows the existing logic in TopicCommand
+              arguments.replicasAssignments.asScala.map { case (partitionId, replicas) =>
+                (partitionId.intValue, replicas.asScala.map(_.intValue).toSeq)
+              }
+            } else {
+              AdminUtils.assignReplicasToBrokers(brokers, arguments.partitions, arguments.replicationFactor)
+            }
+          }
+
+          // Create the topic
+          AdminUtils.createOrUpdateTopicPartitionAssignmentPathInZK(zkUtils, topic, assignments, configs, update = false)
+
+          // Return valid response
+          (topic, Errors.NONE)
+        } catch {
+          // Handle errors with appropriate error response
+          case e: Throwable => (topic, Errors.forException(e))
+        }
+      }
+    }
+
+    val results =
+      if(!authorize(request.session, Create, Resource.ClusterResource)) {
+        handleUnauthorizedRequest
```

Comment:
Thanks for the explanation. I struggled with this a little bit on some of the group APIs. The end result is that there's little consistency between the different handlers. Not sure I see a clear path to improving this though since each request tends to have different needs (the lack of generalizable authorization/error handling makes reuse difficult). Grouping related APIs into separate classes would improve organization, but I think your plan to minimize impact to get KIP-4 through is a good one.